From 627bb9864b9022a3c76cbe7f21ea3f003f13e97b Mon Sep 17 00:00:00 2001
From: Alberto Ponces <ponces26@gmail.com>
Date: Sun, 30 Apr 2023 23:29:04 +0100
Subject: [PATCH 20/24] audiopolicy: Fix broken mic while video recording on
 some Exynos devices

This should fix https://github.com/phhusson/treble_experimentations/issues/2021 and https://github.com/phhusson/treble_experimentations/issues/2384.
Credits to @haridhayal11.
---
 .../common/managerdefinitions/src/Serializer.cpp           | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp b/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
index b9f71cc42b..77a96b47a4 100644
--- a/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
+++ b/services/audiopolicy/common/managerdefinitions/src/Serializer.cpp
@@ -662,6 +662,9 @@ std::variant<status_t, RouteTraits::Element> PolicySerializer::deserialize<Route
     }
     route->setSink(sink);
 
+    // This fixes broken mic while video record on some Exynos devices
+    bool disableBackMic = property_get_bool("persist.sys.phh.disable_back_mic", false);
+
     std::string sourcesAttr = getXmlAttribute(cur, Attributes::sources);
     if (sourcesAttr.empty()) {
         ALOGE("%s: No %s found", __func__, Attributes::sources);
@@ -678,6 +681,10 @@ std::variant<status_t, RouteTraits::Element> PolicySerializer::deserialize<Route
             if (source == NULL) {
                 source = ctx->findPortByTagName(trim(devTag));
             }
+            if (disableBackMic && strcmp(devTag, "Built-In Back Mic") == 0) {
+                ALOGW("Skipping route source \"%s\" as it breaks video recording mic", devTag);
+                source = NULL;
+            }
             if (source == NULL) {
                 if (false && !mIgnoreVendorExtensions) {
                     ALOGE("%s: no source found with name \"%s\"", __func__, devTag);
-- 
2.34.1

